Verify all branches are handled in conditional task in playbooks

DeanArbel opened the pull request:
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/21850

## Description
Added is_condition_branches_handled_correctly to PlaybookValidator, which validates all condition branches are handled correctly.

Itay4 commented on demisto_sdk/commands/common/hook_validations/playbook.py:
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
+                            return False
+                    # if there are task_condition_labels left then not all branches are handled
+                    if task_condition_labels:
+                        print_error('Playbook has conditional task with unhandled branches: {}'.format(
```
use f-strings

Itay4 commented on demisto_sdk/commands/common/hook_validations/playbook.py:
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
+                            if next_task_ids:
+                                task_condition_labels.remove(next_task_branch)
+                        except KeyError as e:
+                            print_error('Playbook has conditional task with unreachable next task: {}'.format(str(e)))
```
use f-strings
what will the error message look like? will it be any indicative?

DeanArbel replied:
Yeah, the str(e) of a key error is the key itself, so it will print the task ID.

Itay4 replied:
please post screenshot of example

DeanArbel replied:
see comment below

DeanArbel commented:
@Itay4 I removed the walrus operator, as it failed the build. demisto-sdk doesn't support py38, and according to @bakatzir, we are not planning on changing that soon.

Itay4 commented:
> @Itay4 I removed the walrus operator, as it failed the build. demisto-sdk doesn't support py38, and according to @bakatzir, we are not planning on changing that soon.

of course we support py3.8... its just the build also runs py3.7
leave it for now, ill handle it later

DeanArbel commented:
@Itay4 following your review I made the error messages more informative, please see the screenshot below as an example:
![image](https://user-images.githubusercontent.com/20818773/76162055-0fdb7d80-6142-11ea-8ffd-67726cf94611.png)

Itay4 commented:
> @Itay4 following your review I made the error messages more informative, please see the screenshot below as an example:
> ![image](https://user-images.githubusercontent.com/20818773/76162055-0fdb7d80-6142-11ea-8ffd-67726cf94611.png)

lets improve the english, something like `... has unhandled condition: <CONDITION-NAME>`
whats `#default` ?

Itay4 commented on demisto_sdk/commands/common/hook_validations/playbook.py:
```diff
@@ -21,7 +23,8 @@ def is_valid_playbook(self, is_new_playbook=True):  # type: (bool) -> bool
                 self.is_no_rolename(),
-                self.is_root_connected_to_all_tasks()
+                self.is_root_connected_to_all_tasks(),
+                self.is_condition_branches_handled_correctly()
```
```suggestion
                self.is_all_condition_branches_handled()
```

Itay4 commented on demisto_sdk/commands/common/hook_validations/playbook.py:
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
             return False
         return True
+
+    def is_condition_branches_handled_correctly(self):  # type: () -> bool
```
add return type hint

Itay4 commented on demisto_sdk/commands/common/hook_validations/playbook.py:
```diff
@@ -65,6 +69,38 @@ def is_no_rolename(self):  # type: () -> bool
+        Return:
+            bool. if the Playbook handles all condition branches correctly.
+        """
+        is_unreachable: bool = True
```
the variable name here is misleading, lets improve it

Itay4 commented on demisto_sdk/commands/common/tests/playbook_test.py:
```diff
@@ -75,6 +121,18 @@ def test_is_added_required_fields(self, current_file, answer):
+        validator = PlaybookValidator(structure)
+        assert validator.is_no_rolename() is answer
+
+    @pytest.mark.parametrize("current_file, answer", IS_CONDITIONAL_INPUTS)
```
why do we name these `current_file` and `answer` ? this is confusing..

DeanArbel replied:
changed to `playbook_json` and `expected_result`

Itay4 commented on demisto_sdk/tests/test_files/playbook-Invalid_condition_1.yml:
```diff
@@ -0,0 +1,111 @@
+elasticcommonfields: {}
+id: Invalid_condition_1
```
for both playbooks with invalid conditions, lets write why they're invalid

DeanArbel replied:
changed to `Invalid_condition_unhandled_branch` and `Invalid_condition_unhandled_branch_and_unhandled_condition`

DeanArbel commented:
@Itay4 ran the validation on all content - didn't encounter an error with the playbooks.

Itay4 commented:
@DeanArbel see my comment https://github.com/demisto/demisto-sdk/pull/239#issuecomment-596225551
the first error message is not clear enough, we can improve it
what is `default`? what is the issue in this case?

DeanArbel commented:
> @DeanArbel see my comment [#239 (comment)](https://github.com/demisto/demisto-sdk/pull/239#issuecomment-596225551)
> the first error message is not clear enough, we can improve it
> what is `default`? what is the issue in this case?

1. Changed it, please see screenshot.
2. `#default#` is the `else` condition. I removed the # chars, and replaced default with else. Please see screenshot.

![image](https://user-images.githubusercontent.com/20818773/76224948-8650ac00-6224-11ea-857e-18177e048a5d.png)
